feat: Add css transitions

JayaKrishnaNamburu opened the pull request:
## Description

Adds support for `css-transitions` when users are in `local` state. In any other state than local, the CRUD operations on transitions are disabled.

JayaKrishnaNamburu commented:
@taylornowotny @kof the transitions section is in usable state. The UX is very similar to `box-shadow`. So give it a try and see the changes needed.

kof commented:
@JayaKrishnaNamburu testing

kof commented:
- [ ] I think it makes sense to have WF defaults when adding transition:

<img width="249" alt="Screenshot 2023-10-22 at 10 11 09" src="https://github.com/webstudio-is/webstudio/assets/52824/01b01d56-06a1-44fb-895d-a08400c9c13d">

kof commented:
Is the feature still behind a feature flag? because we can't merge if user will see it as-is right away

istarkov commented on apps/builder/app/builder/features/style-panel/shared/use-intermediate-layer.ts:
```diff
@@ -0,0 +1,46 @@
+  value: string;
+};
+
+export const useIntermediateContent = ({
```
First this hook is written with bug. It copypasted from buggy version. And this bug are in few places.
intermediate value should be used only during edit. And should be reseted after onComplete. Otherwise any changes to style from other places (like AI changes) will never be shown.

2nd I don't think that this hook need to be shared. It has no value to be shareable. I don't even like its a hook, as its just handlers and state.

istarkov commented on packages/css-data/bin/mdn-data.ts:
```diff
@@ -233,6 +233,22 @@ const filteredProperties: FilteredProperties = (() => {
     const isSupportedStatus =
       config.status === "standard" || config.status === "experimental";
+
+    /*
```
use GPT to fix english errors ;)  prperties -> .... etc

istarkov replied:
Also I don;t understand comment, example will be better

istarkov commented on packages/css-data/bin/mdn-data.ts:
```diff
@@ -233,6 +233,22 @@ const filteredProperties: FilteredProperties = (() => {
+
+    */
+    if (property === "transition") {
+      config.initial = "all 0s ease 0s";
```
Why we use config initial here, what it solves?

istarkov commented on packages/css-data/src/property-parsers/transition.ts:
```diff
@@ -0,0 +1,105 @@
+  }
+};
+
+export const parseTransition = (
```
useless without tests no one will get what this function do

kof replied:
definietely needs tests like other similar functions have

JayaKrishnaNamburu replied:
Yeah, completely forgot adding them when opening for review. Will take a look at all the comments and the PR once again 👍

istarkov commented on packages/css-data/src/property-parsers/transition.ts:
```diff
@@ -0,0 +1,105 @@
+
+  const layers: TupleValue[] = [];
+
+  csstree.walk(cssAst, (node) => {
```
Imo underlying algorithm should be simple - simple.
It's just splitting children array on Operator `,` but too much logic for this simple thing.
like this comparisons `children.last === child` looks useless

JayaKrishnaNamburu replied:
The first iteration of design has individual display like, so the parser was little complicated to pick everything. Looks like the design is changed now, and the parser also can be just simple one now.

taylornowotny reviewed (changes requested):
Tested and works for me!

UI feedback:
- (bug) click an item to open the menu, then close the menu. You'll see the "remove" button stays visible when it should be hidden.
- There should be no space between items.

JayaKrishnaNamburu commented:
Fixes the `disabled` state @taylornowotny. But regarding the second one

> There should be no space between items.
It's just the `CSSValueListItem` component. So, if it's changed here. It reflects everywhere. I am not sure, if need to be changed everywhere. Can you confirm on this once.

And regarding rest of all the comments, @istarkov handled a bunch of them. Added tests and added new comments in the `mdn-data` file for the `config.initial` changes

taylornowotny commented:
Yes Jaya, all CSSValueListItems should not have spaces between items.
Thanks!

On Tue, Oct 24, 2023 at 3:04 AM Jaya Krishna Namburu <
***@***.***> wrote:

> Fixes the disabled state @taylornowotny <https://github.com/taylornowotny>.
> But regarding the second one
>
> There should be no space between items.
> It's just the CSSValueListItem component. So, if it's changed here. It
> reflects everywhere. I am not sure, if need to be changed everywhere. Can
> you confirm on this once.
>
> And regarding rest of all the comments, @istarkov
> <https://github.com/istarkov> handled a bunch of them. Added tests and
> added new comments in the mdn-data file for the config.initial changes
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/webstudio-is/webstudio/pull/2511#issuecomment-1776906709>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUT5FGF2IIXZLCSAJ26JBDDYA6HDJAVCNFSM6AAAAAA6ISIKT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZWHEYDMNZQHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JayaKrishnaNamburu commented:
Done @taylornowotny

taylornowotny commented:
Thanks @JayaKrishnaNamburu space is good now.

Here's a video of that bug I mentioned in case it wasn't clear

https://github.com/webstudio-is/webstudio/assets/86495896/ec899303-e90b-4697-b469-6c7b52dab983

istarkov commented on apps/builder/app/builder/features/style-panel/sections/transitions/transition-content.tsx:
```diff
@@ -0,0 +1,75 @@
+      return;
+    }
+
+    props.onEditLayer(props.index, layers);
```
I think we need to reset intermediate value to undefined in handleComplete. Imagine you have edited it, then AI will change store value, but in control you will always see the local state

istarkov reviewed (changes requested):
Plz see my comments

taylornowotny commented:
Floating panel UI is done and documented!

Please read the documentation very carefully to see all the details.

https://github.com/webstudio-is/webstudio/issues/1939#issuecomment-1778141375

JayaKrishnaNamburu commented:
Please don't do any code / ux reviews. I started with the UX. So, takes some time.

taylornowotny reviewed (changes requested):
Transitions panel UI not yet implemented

Remove button bug that I pointed out is still there

JayaKrishnaNamburu commented:
@taylornowotny i just updated the upstream. Did't push any chaanges yet. Will let you know

JayaKrishnaNamburu commented:
And @taylornowotny i just tested it. It's not bug, but more of a UX thing maybe ?
Because once the floating panel is closed. The focus is moved back to the element that is just being edited. So, the selector is kicking and marking them as visible. 

https://github.com/webstudio-is/webstudio/blob/main/packages/design-system/src/components/css-value-list-item.tsx#L121

https://github.com/webstudio-is/webstudio/assets/11075561/7b6113c4-6896-4540-8d5d-3bd57009c71c

taylornowotny commented:
@JayaKrishnaNamburu 

After the menu closes, the item should not be focused.

Does that answer your question?

kof commented:
@taylornowotny I just checked, its the same behavior in production, so this pr didn't change it, I would suggest always try on main branch to see if the change was implemented here.

What seems wrong about the current state is not the focus behavior itself, but the outline. When closing panel, the item is still focused, but there is no outline, that makes it seem like its not focused but it is and icons are also still visible.

Do you think we should fix the outline and make it visible after closing? focus stays on the item in that case as it was before panel was opened. So user can keep navigating using arrow keys from where they left

JayaKrishnaNamburu commented:
@taylornowotny we moved the bug to a new issue here. And will be handled separately. https://github.com/webstudio-is/webstudio/issues/2543
Let's move the discussion on this bug there, so all the discussions will be in one place there 👍

taylornowotny reviewed (approved):
Bug moved to a separate issue.

UI not implemented - I'm creating a separate issue for this. #2546 

Approving just so I'm not blocking this.
